svd-encoder/src/peripheral.rs

```diff
@@ -102,10 +104,41 @@ impl Encode for PeripheralInfo {
         elem.children.append(&mut interrupts?);
 
         if let Some(v) = &self.registers {
-            let children: Result<Vec<_>, _> = v
-                .iter()
-                .map(|e| e.encode_node_with_config(config))
-                .collect();
+            let children: Result<Vec<_>, _> =
+                if let Some(first) = config.registers_or_clusters_first {
+                    let mut reg_refs = v
+                        .iter()
+                        .filter(|rc| matches!(rc, RegisterCluster::Register(_)))
+                        .collect::<Vec<_>>();
+                    sort_register_cluster(&mut reg_refs, config.register_cluster_sorting);
+                    let mut c_refs = v
+                        .iter()
+                        .filter(|rc| matches!(rc, RegisterCluster::Cluster(_)))
+                        .collect::<Vec<_>>();
+                    sort_register_cluster(&mut c_refs, config.register_cluster_sorting);
+                    match first {
+                        crate::RegistersOrClustersFirst::Registers => {
+                            reg_refs.into_iter().chain(c_refs.into_iter())
+                        }
+                        crate::RegistersOrClustersFirst::Clusters => {
+                            c_refs.into_iter().chain(reg_refs.into_iter())
+                        }
+                    }
+                    .map(|e| e.encode_node_with_config(config))
+                    .collect()
+                } else {
+                    if let Some(sorting) = config.register_cluster_sorting {
+                        let mut refs = v.iter().collect::<Vec<_>>();
+                        sort_register_cluster(&mut refs, Some(sorting));
+                        refs.into_iter()
+                            .map(|e| e.encode_node_with_config(config))
+                            .collect()
+                    } else {
+                        v.iter()
+                            .map(|e| e.encode_node_with_config(config))
+                            .collect()
+                    }
+                };
```

Comment:
clippy wants to merge these, I think I agree

Reply:
Second branch is default. It should be fastest and avoid advanced allocations.